bpf/kmesh/workload/include/backend.h

```diff
@@ -15,49 +15,58 @@ static inline backend_value *map_lookup_backend(const backend_key *key)
     return kmesh_map_lookup_elem(&map_of_backend, key);
 }
 
-static inline int waypoint_manager(ctx_buff_t *ctx, __u32 addr, __u32 port)
+static inline int waypoint_manager(struct kmesh_context *kmesh_ctx, struct ip_addr *wp_addr, __u32 port)
 {
     int ret;
     address_t target_addr;
+    ctx_buff_t *ctx = (ctx_buff_t *)kmesh_ctx->ctx;
     __u64 *sk = (__u64 *)ctx->sk;
     struct bpf_sock_tuple value_tuple = {0};
 
-    value_tuple.ipv4.daddr = ctx->user_ip4;
-    value_tuple.ipv4.dport = ctx->user_port;
-
-    ret = bpf_map_update_elem(&map_of_dst_info, &sk, &value_tuple, BPF_NOEXIST);
+    if (ctx->family == AF_INET) {
+        value_tuple.ipv4.daddr = kmesh_ctx->vip.ip4;
+        value_tuple.ipv4.dport = ctx->user_port;
+    } else if (ctx->family == AF_INET6) {
+        bpf_memcpy(value_tuple.ipv6.daddr, kmesh_ctx->vip.ip6, IPV6_ADDR_LEN);
+        value_tuple.ipv6.dport = ctx->user_port;
+    } else {
+        BPF_LOG(ERR, BACKEND, "invalid ctx family: %u\n", ctx->family);
+        return -1;
+    }
+    ret = bpf_map_update_elem(&map_of_dst_info, &(sk), &value_tuple, BPF_NOEXIST);
     if (ret) {
         BPF_LOG(ERR, BACKEND, "record metadata origin address and port failed, ret is %d\n", ret);
         return ret;
     }
-    target_addr.ipv4 = addr;
-    target_addr.port = port;
-    SET_CTX_ADDRESS(ctx, target_addr);
-    kmesh_workload_tail_call(ctx, TAIL_CALL_CONNECT4_INDEX);
 
-    // if tail call failed will run this code
-    BPF_LOG(ERR, BACKEND, "workload tail call failed, err is %d\n", ret);
-    return -ENOEXEC;
+    if (ctx->user_family == AF_INET)
+        kmesh_ctx->dnat_ip.ip4 = wp_addr->ip4;
+    else
+        bpf_memcpy(kmesh_ctx->dnat_ip.ip6, wp_addr->ip6, IPV6_ADDR_LEN);
+    kmesh_ctx->dnat_port = port;
+    kmesh_ctx->via_waypoint = true;
```

Comment:
if we move this tail call into this function, it seems we donot need `via_waypoint`

Reply:
Before tail call, need dnat dstIp(v4/v6), bpf verifier will fail.